src/simsopt/field/force.py

```diff
@@ -182,92 +442,2193 @@ def mean_squared_force_pure(gamma, gammadash, gammadashdash, quadpoints, current
+    C = jnp.linalg.cholesky(coil_coil_inductances_full_pure(gammas, gammadashs, a_list, b_list, downsample, cross_section))
+    inv_C = jscp.linalg.solve_triangular(C, jnp.eye(C.shape[0]), lower=True)
+    inv_L = jscp.linalg.solve_triangular(C.T, inv_C, lower=False)
+    return inv_L
+
+
+def coil_currents_barebones(gammas, gammadashs, gammas_TF, gammadashs_TF, currents_TF, a_list, b_list, downsample, cross_section):
+    """
+    Pure function for computing the induced currents in a set of passive coils 
+    due to a set of TF coils + the other passive coils. 
+
+    Args:
+        gammas (array): Array of passive coil positions.
+        gammadashs (array): Array of passive coil tangent vectors.
+        gammas_TF (array): Array of TF coil positions.
+        gammadashs_TF (array): Array of TF coil tangent vectors.
+        currents_TF (array): Array of TF coil current.
+        a_list (array): Array of passive coil radii.
+        b_list (array): Array of passive coil thicknesses.
+        downsample (int): Downsample factor for the objective function.
+        cross_section (str): Cross-section type.
+
+    Returns:
+        array: Array of induced currents.
+    """
+    return -coil_coil_inductances_inv_pure(gammas, gammadashs, a_list, b_list, downsample, cross_section) @ net_fluxes_pure(gammas, gammadashs, gammas_TF, gammadashs_TF, currents_TF, downsample)
+
+
+def tve_pure(gamma, gammadash, gammas2, gammadashs2, current, currents2, a, b, downsample, cross_section):
+    r"""Pure function for minimizing the total vacuum energy on a coil.
+
+    The function is
+
+     .. math::
+        J = \frac{1}{2}I_iL_{ij}I_j
+
+    where :math:`L_{ij}` is the coil inductance matrix (positive definite),
+    and :math:`I_i` is the current in the ith coil.
+
+    Args:
+        gamma (array): Array of coil positions.
+        gammadash (array): Array of coil tangent vectors.
+        gammas2 (array): Array of coil positions.
+        gammadashs2 (array): Array of coil tangent vectors.
+        current (array): Array of coil current.
+        currents2 (array): Array of coil current.
+        a (float): Coil radius.
+        b (float): Coil thickness.
+        downsample (int): Downsample factor for the objective function.
+        cross_section (str): Cross-section type.   
+
+    Returns:
+        float: Value of the objective function.
+    """
+    Ii_Ij = (current * currents2)
+    Lij = coil_coil_inductances_pure(
+        gamma, gammadash, gammas2, gammadashs2, a, b, downsample, cross_section
+    )
+    U = 0.5 * (jnp.sum(Ii_Ij * Lij[1:]) + Lij[0] * current ** 2)
+    return U
+
+
+class TVE(Optimizable):
+    r"""Optimizable class for minimizing the total vacuum energy on a coil.
+
+    The function is
+
+     .. math::
+        J = \frac{1}{2}I_iL_{ij}I_j
+
+    where :math:`L_{ij}` is the coil inductance matrix (positive definite),
+    and :math:`I_i` is the current in the ith coil.
+
+    Args:
+        coil (Coil): Coil to optimize.
```

Comment:
It seems unnecessarily complicated that the TVE class forces the user to separate out one coil from the rest, and then manually sum over a list of TVE objects, as in line 183 of `coil_force_objectives_scan.py`: `sum([TVE(c, coils, a=a) for c in base_coils])`. The original formula for the energy is symmetric, $W = \sum_{i,j} L_{ij} I_i I_j$, so why not have this class just accept the list of all coils, and treat everything in a symmetric way? The API would be simpler then: `J = TVE(coils, a=a)`.